// @tosspayments/widget-sdk-react-native/lib/module/components/PaymentMethodWidget.js
import { useEffect, useRef, useState } from 'react';
import { View } from 'react-native';
import { defaultInjectedJavascript, htmlString } from '../documents/html';
import { usePaymentWidgetContext } from '../hooks/usePaymentWidget';
import { useWidgetRegistry } from '../hooks/useWidgetRegistry';
import { PaymentsWebView } from '../webview/PaymentsWebView';
import { RequestPaymentsModal } from './RequestPaymentsModal';
export function PaymentMethodWidget(_ref) {
  var _context$options;
  let {
    selector,
    onCustomRequest,
    onCustomPaymentMethodSelect,
    onCustomPaymentMethodUnselect,
    onLoadEnd
  } = _ref;
  const [height, setHeight] = useState(400);
  const webview = useRef(null);
  const context = usePaymentWidgetContext();
  const {
    register,
    unregister
  } = useWidgetRegistry();
  const [isModalVisible, setModalVisible] = useState(false);
  const [requestPaymentHTML, setRequestPaymentHTML] = useState(null);
  useEffect(() => {
    register(selector, webview);
    return () => {
      unregister(selector);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return /*#__PURE__*/React.createElement(View, {
    style: {
      height,
      width: '100%'
    }
  }, /*#__PURE__*/React.createElement(RequestPaymentsModal, {
    requestPaymentHTML: requestPaymentHTML,
    isVisible: isModalVisible,
    onEvaluateJavascript: script => {
      var _webview$current;
      (_webview$current = webview.current) === null || _webview$current === void 0 || _webview$current.injectJavaScript(script);
      setModalVisible(false);
    },
    onSuccess: success => {
      var _webview$current2, _webview$current2$suc;
      (_webview$current2 = webview.current) === null || _webview$current2 === void 0 || (_webview$current2$suc = _webview$current2.success) === null || _webview$current2$suc === void 0 || _webview$current2$suc.call(_webview$current2, success);
      setModalVisible(false);
    },
    onFail: fail => {
      var _webview$current3, _webview$current3$fai;
      (_webview$current3 = webview.current) === null || _webview$current3 === void 0 || (_webview$current3$fai = _webview$current3.fail) === null || _webview$current3$fai === void 0 || _webview$current3$fai.call(_webview$current3, fail);
      setModalVisible(false);
    }
  }), /*#__PURE__*/React.createElement(PaymentsWebView, {
    androidLayerType: 'none',
    webviewDebuggingEnabled: true,
    ref: webview,
    source: {
      html: htmlString({
        stage: 'v1',
        selector
      }),
      baseUrl: ((_context$options = context.options) === null || _context$options === void 0 || (_context$options = _context$options.brandpay) === null || _context$options === void 0 ? void 0 : _context$options.redirectUrl) ?? 'https://tosspayments.com'
    },
    injectedJavaScript: defaultInjectedJavascript,
    onLoadEnd: onLoadEnd,
    onMessage: event => {
      var _webview$current4;
      if (event.nativeEvent.data !== undefined) {
        const jsonObject = JSON.parse(event.nativeEvent.data);
        switch (jsonObject.name) {
          case 'updateHeight':
            setHeight(jsonObject.params.height);
            break;
          case 'requestPayments':
          case 'requestHTML':
            setRequestPaymentHTML(jsonObject.params.html);
            setModalVisible(true);
            break;
          case 'widgetStatus':
            // TODO
            break;
          case 'error':
            // TODO
            break;
          case 'customRequest':
            onCustomRequest === null || onCustomRequest === void 0 || onCustomRequest({
              paymentMethodKey: jsonObject.params.paymentMethodKey
            });
            break;
          case 'customPaymentMethodSelect':
            onCustomPaymentMethodSelect === null || onCustomPaymentMethodSelect === void 0 || onCustomPaymentMethodSelect({
              paymentMethodKey: jsonObject.params.paymentMethodKey
            });
            break;
          case 'customPaymentMethodUnselect':
            onCustomPaymentMethodUnselect === null || onCustomPaymentMethodUnselect === void 0 || onCustomPaymentMethodUnselect({
              paymentMethodKey: jsonObject.params.paymentMethodKey
            });
            break;
          default:
            (_webview$current4 = webview.current) === null || _webview$current4 === void 0 || _webview$current4.onMessage(event);
            break;
        }
      }
    },
    containerStyle: {
      height
    }
  }));
}
