// @tosspayments/widget-sdk-react-native/src/webview/PaymentsWebView.tsx
import React, { Component } from 'react';
import { Alert, Linking } from 'react-native';
import WebView, { type WebViewProps } from 'react-native-webview';
import type { WebViewSharedProps } from 'react-native-webview/lib/WebViewTypes';
import type { Fail } from '../models/Result';

type PaymentsWebViewProps = WebViewSharedProps & WebViewProps;

export class PaymentsWebView extends Component<PaymentsWebViewProps> {
  webRef: React.MutableRefObject<WebView | null>;
  resolveMap: Record<string, (resolve: any) => void>;
  rejectMap: Record<string, (reject: any) => void>;
  renderSuccess?: () => void;
  renderFail?: (fail: Fail) => void;

  constructor(props: PaymentsWebViewProps) {
    super(props);
    this.resolveMap = {};
    this.rejectMap = {};
    this.webRef = React.createRef();
  }

  injectJavaScript: (javascript: string) => void = (javascript) => {
    this.webRef.current?.injectJavaScript(javascript);
  };

  onMessage: (message: any) => void = (message) => {
    if (message.nativeEvent.data !== undefined) {
      const jsonObject = JSON.parse(message.nativeEvent.data);
      switch (jsonObject.name) {
        case 'resolve':
          const resolve = this.resolveMap[jsonObject.params.key];
          resolve?.(jsonObject.params.data);
          delete this.resolveMap[jsonObject.params.key];
          delete this.rejectMap[jsonObject.params.key];
          break;
        case 'reject':
          const reject = this.rejectMap[jsonObject.params.key];
          reject?.(jsonObject.params.data);
          delete this.resolveMap[jsonObject.params.key];
          delete this.rejectMap[jsonObject.params.key];
          break;
        case 'changePaymentMethod':
          // do nothing;
          break;
        default:
          Alert.alert(`Widget에 ${jsonObject.name} 을 구현해주세요`);
          break;
      }
    }
  };

  /**
   * NOTE(@genie-youn): URL로 판단하지 않고 URL을 디바이스 브라우저로 열라고 명시적으로 메세지로 요청하도록 변경필요
   */
  render() {
    return (
      <WebView
        ref={this.webRef}
        style={{
          // NOTE(JooYang): hardwareAccelerated 크래시 해결
          opacity: 0.99,
        }}
        onShouldStartLoadWithRequest={(event) => {
          if (
            event.mainDocumentURL != null &&
            this.isRequiredOpenInDeviceBrowser(event.mainDocumentURL)
          ) {
            Linking.openURL(event.mainDocumentURL);

            return false;
          }

          return true;
        }}
        {...this.props}
      />
    );
  }

  /**
   * asyncEvaluateJavascript
   */
  async asyncEvaluateJavascript(
    script: string,
    randomKey: string = ''
  ): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      this.resolveMap[randomKey] = (data) => {
        resolve(JSON.stringify(data));
      };
      this.rejectMap[randomKey] = (data) => {
        reject(JSON.stringify(data));
      };
      this.webRef.current?.injectJavaScript(script);
    });
  }

  private isRequiredOpenInDeviceBrowser(url: string): boolean {
    const targetUrl = [
      'https://pgweb.tosspayments.com/tosspayments/MainPopUp.do',
      'https://pages.tosspayments.com/terms/privacy/consent_address',
      'https://pages.tosspayments.com/terms/privacy/consent_provide_shoppingpay',
      'https://pages.tosspayments.com/terms/user',
      'https://pages.tosspayments.com/terms/privacy/consent1_new',
      'https://pages.tosspayments.com/terms/privacy/consent2_new',
    ];

    return targetUrl.includes(url);
  }
}
