// @tosspayments/widget-sdk-react-native/lib/commonjs/components/PaymentMethodWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaymentMethodWidget = PaymentMethodWidget;
var _react = require("react");
var _reactNative = require("react-native");
var _html = require("../documents/html");
var _usePaymentWidget = require("../hooks/usePaymentWidget");
var _useWidgetRegistry = require("../hooks/useWidgetRegistry");
var _RequestPaymentsModal = require("./RequestPaymentsModal");
var _PaymentMethodWebView = require("../webview/PaymentMethodWebView");
function PaymentMethodWidget(_ref) {
  var _context$options;
  let {
    selector,
    onCustomRequest,
    onCustomPaymentMethodSelect,
    onCustomPaymentMethodUnselect,
    onLoadEnd
  } = _ref;
  const [height, setHeight] = (0, _react.useState)(400);
  const webview = (0, _react.useRef)(null);
  const context = (0, _usePaymentWidget.usePaymentWidgetContext)();
  const {
    register,
    unregister
  } = (0, _useWidgetRegistry.useWidgetRegistry)();
  const [isModalVisible, setModalVisible] = (0, _react.useState)(false);
  const [requestPaymentHTML, setRequestPaymentHTML] = (0, _react.useState)(null);
  (0, _react.useEffect)(() => {
    register(selector, webview);
    return () => {
      unregister(selector);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return /*#__PURE__*/React.createElement(_reactNative.View, {
    style: {
      height,
      width: '100%'
    }
  }, /*#__PURE__*/React.createElement(_RequestPaymentsModal.RequestPaymentsModal, {
    requestPaymentHTML: requestPaymentHTML,
    isVisible: isModalVisible,
    onEvaluateJavascript: script => {
      var _webview$current;
      (_webview$current = webview.current) === null || _webview$current === void 0 || _webview$current.injectJavaScript(script);
      setModalVisible(false);
    },
    onSuccess: success => {
      var _webview$current2, _webview$current2$pay;
      (_webview$current2 = webview.current) === null || _webview$current2 === void 0 || (_webview$current2$pay = _webview$current2.paymentSuccess) === null || _webview$current2$pay === void 0 || _webview$current2$pay.call(_webview$current2, success);
      setModalVisible(false);
    },
    onFail: fail => {
      var _webview$current3, _webview$current3$pay;
      (_webview$current3 = webview.current) === null || _webview$current3 === void 0 || (_webview$current3$pay = _webview$current3.paymentFail) === null || _webview$current3$pay === void 0 || _webview$current3$pay.call(_webview$current3, fail);
      setModalVisible(false);
    }
  }), /*#__PURE__*/React.createElement(_PaymentMethodWebView.PaymentMethodWebView, {
    androidLayerType: 'none',
    webviewDebuggingEnabled: true,
    ref: webview,
    source: {
      html: (0, _html.htmlString)({
        stage: 'v1',
        selector
      }),
      baseUrl: ((_context$options = context.options) === null || _context$options === void 0 || (_context$options = _context$options.brandpay) === null || _context$options === void 0 ? void 0 : _context$options.redirectUrl) ?? 'https://tosspayments.com'
    },
    injectedJavaScript: _html.defaultInjectedJavascript,
    onLoadEnd: onLoadEnd,
    onMessage: event => {
      var _webview$current5, _webview$current5$ren, _webview$current6, _webview$current6$pay, _webview$current7;
      if (event.nativeEvent.data !== undefined) {
        const jsonObject = JSON.parse(event.nativeEvent.data);
        switch (jsonObject.name) {
          case 'updateHeight':
            setHeight(jsonObject.params.height);
            break;
          case 'requestPayments':
          case 'requestHTML':
            setRequestPaymentHTML(jsonObject.params.html);
            setModalVisible(true);
            break;
          case 'widgetStatus':
            const params = jsonObject.params;
            if (params.status === 'load' && params.widget === 'paymentMethods') {
              var _webview$current4, _webview$current4$ren;
              webview === null || webview === void 0 || (_webview$current4 = webview.current) === null || _webview$current4 === void 0 || (_webview$current4$ren = _webview$current4.renderSuccess) === null || _webview$current4$ren === void 0 || _webview$current4$ren.call(_webview$current4);
            }
            break;
          case 'error':
            // FIXME : 논의 필요
            webview === null || webview === void 0 || (_webview$current5 = webview.current) === null || _webview$current5 === void 0 || (_webview$current5$ren = _webview$current5.renderFail) === null || _webview$current5$ren === void 0 || _webview$current5$ren.call(_webview$current5, {
              code: jsonObject.params.errorCode,
              message: jsonObject.params.errorMessage,
              orderId: 'unknown'
            });
            webview === null || webview === void 0 || (_webview$current6 = webview.current) === null || _webview$current6 === void 0 || (_webview$current6$pay = _webview$current6.paymentFail) === null || _webview$current6$pay === void 0 || _webview$current6$pay.call(_webview$current6, {
              code: jsonObject.params.errorCode,
              message: jsonObject.params.errorMessage,
              orderId: 'unknown'
            });
            break;
          case 'customRequest':
            onCustomRequest === null || onCustomRequest === void 0 || onCustomRequest({
              paymentMethodKey: jsonObject.params.paymentMethodKey
            });
            break;
          case 'customPaymentMethodSelect':
            onCustomPaymentMethodSelect === null || onCustomPaymentMethodSelect === void 0 || onCustomPaymentMethodSelect({
              paymentMethodKey: jsonObject.params.paymentMethodKey
            });
            break;
          case 'customPaymentMethodUnselect':
            onCustomPaymentMethodUnselect === null || onCustomPaymentMethodUnselect === void 0 || onCustomPaymentMethodUnselect({
              paymentMethodKey: jsonObject.params.paymentMethodKey
            });
            break;
          default:
            (_webview$current7 = webview.current) === null || _webview$current7 === void 0 || _webview$current7.onMessage(event);
            break;
        }
      }
    },
    containerStyle: {
      height
    }
  }));
}
