// @tosspayments/widget-sdk-react-native/lib/commonjs/classes/PaymentWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaymentWidget = void 0;
var _AgreementWidgetControl = require("./AgreementWidgetControl");
var _PaymentMethodWidgetControl = require("./PaymentMethodWidgetControl");
class PaymentWidget {
  constructor(clientKey, customerKey, options) {
    this.clientKey = clientKey;
    this.customerKey = customerKey;
    this.options = options;
  }
  equals(paymentWidget) {
    if (paymentWidget == null) {
      return false;
    }
    return paymentWidget.clientKey === this.clientKey && paymentWidget.customerKey === this.customerKey && JSON.stringify(paymentWidget.options) === JSON.stringify(this.options);
  }
  async renderPaymentMethods(ref, selector, amount, options) {
    this.amount = amount;
    const methodWidgetRef = ref.current;
    const widgetOptions = this.options ?? {};
    methodWidgetRef === null || methodWidgetRef === void 0 || methodWidgetRef.injectJavaScript(`
    const widget = PaymentWidget('${this.clientKey}', '${this.customerKey}', ${JSON.stringify(widgetOptions)});
    const paymentMethodWidget = widget.renderPaymentMethods('#${selector}', ${JSON.stringify(amount)}, { variantKey: ${(options === null || options === void 0 ? void 0 : options.variantKey) !== undefined ? `'${options === null || options === void 0 ? void 0 : options.variantKey}'` : `undefined`} });
    true;
    `);

    //TODO: force-unwraping
    return new _PaymentMethodWidgetControl.PaymentMethodWidgetControl(ref);
  }
  async renderAgeements(ref, selector, options) {
    const agreementWidgetRef = ref === null || ref === void 0 ? void 0 : ref.current;
    const widgetOptions = this.options ?? {};
    const widgetScript = `
    const widget = PaymentWidget(
      '${this.clientKey}',
      '${this.customerKey}',
      ${JSON.stringify(widgetOptions)}
    );
    `;
    const renderScript = `${(options === null || options === void 0 ? void 0 : options.variantKey) !== undefined ? `const agreementWidget = widget.renderAgreement('#${selector}', ${JSON.stringify(options)});` : `const agreementWidget = widget.renderAgreement('#${selector}');`}`;
    agreementWidgetRef === null || agreementWidgetRef === void 0 || agreementWidgetRef.injectJavaScript(`
    ${widgetScript}
    ${renderScript}
    true;
    `);

    //TODO: force-unwraping
    return new _AgreementWidgetControl.AgreementWidgetControl(ref);
  }
  async requestPayments(ref, info) {
    const mainWidgetRef = ref === null || ref === void 0 ? void 0 : ref.current;
    if (mainWidgetRef == null) {
      return Promise.reject('mainWidget is undefined');
    }
    const anyInfo = info;
    anyInfo.successUrl = 'tosspayments://success';
    anyInfo.failUrl = 'tosspayments://fail';
    anyInfo.amount = this.amount;
    return mainWidgetRef.requestPayments(anyInfo);
  }
}
exports.PaymentWidget = PaymentWidget;
